// tus-js-client/lib.esm/node/urlStorage.ts
// @ts-expect-error TS(7023): '_typeof' implicitly has return type 'any' because... Remove this comment to see the full error message
function _typeof(o: any) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o: any) { return typeof o; } : function (o: any) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(instance: any, Constructor: any) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target: any, props: any) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor: any, protoProps: any, staticProps: any) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg: any) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input: any, hint: any) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
/* eslint no-unused-vars: 0 */
import { readFile, writeFile } from 'fs';
// @ts-expect-error TS(7016): Could not find a declaration file for module 'prop... Remove this comment to see the full error message
import * as lockfile from 'proper-lockfile';
// @ts-expect-error TS(7016): Could not find a declaration file for module 'comb... Remove this comment to see the full error message
import combineErrors from 'combine-errors';
export var canStoreURLs = true;
export var FileUrlStorage = /*#__PURE__*/function () {
  function FileUrlStorage(filePath: any) {
    _classCallCheck(this, FileUrlStorage);
    this.path = filePath;
  }
  // @ts-expect-error TS(2554): Expected 3 arguments, but got 2.
  _createClass(FileUrlStorage, [{
    key: "findAllUploads",
    value: function findAllUploads() {
      var _this = this;
      return new Promise(function (resolve, reject) {
        _this._getItems('tus::', function (err: any, results: any) {
          if (err) reject(err);else resolve(results);
        });
      });
    }
  }, {
    key: "findUploadsByFingerprint",
    value: function findUploadsByFingerprint(fingerprint: any) {
      var _this2 = this;
      return new Promise(function (resolve, reject) {
        _this2._getItems("tus::".concat(fingerprint), function (err: any, results: any) {
          if (err) reject(err);else resolve(results);
        });
      });
    }
  }, {
    key: "removeUpload",
    value: function removeUpload(urlStorageKey: any) {
      var _this3 = this;
      return new Promise(function (resolve, reject) {
        _this3._removeItem(urlStorageKey, function (err: any) {
          // @ts-expect-error TS(2794): Expected 1 arguments, but got 0. Did you forget to... Remove this comment to see the full error message
          if (err) reject(err);else resolve();
        });
      });
    }
  }, {
    key: "addUpload",
    value: function addUpload(fingerprint: any, upload: any) {
      var _this4 = this;
      var id = Math.round(Math.random() * 1e12);
      // @ts-expect-error TS(2345): Argument of type 'number' is not assignable to par... Remove this comment to see the full error message
      var key = "tus::".concat(fingerprint, "::").concat(id);
      return new Promise(function (resolve, reject) {
        _this4._setItem(key, upload, function (err: any) {
          if (err) reject(err);else resolve(key);
        });
      });
    }
  }, {
    key: "_setItem",
    value: function _setItem(key: any, value: any, cb: any) {
      var _this5 = this;
      lockfile.lock(this.path, this._lockfileOptions()).then(function (release: any) {
        cb = _this5._releaseAndCb(release, cb);
        _this5._getData(function (err: any, data: any) {
          if (err) {
            cb(err);
            return;
          }
          data[key] = value;
          _this5._writeData(data, function (err2: any) {
            return cb(err2);
          });
        });
      })["catch"](cb);
    }
  }, {
    key: "_getItems",
    value: function _getItems(prefix: any, cb: any) {
      this._getData(function (err: any, data: any) {
        if (err) {
          cb(err);
          return;
        }
        var results = Object.keys(data).filter(function (key) {
          return key.startsWith(prefix);
        }).map(function (key) {
          var obj = data[key];
          obj.urlStorageKey = key;
          return obj;
        });
        cb(null, results);
      });
    }
  }, {
    key: "_removeItem",
    value: function _removeItem(key: any, cb: any) {
      var _this6 = this;
      lockfile.lock(this.path, this._lockfileOptions()).then(function (release: any) {
        cb = _this6._releaseAndCb(release, cb);
        _this6._getData(function (err: any, data: any) {
          if (err) {
            cb(err);
            return;
          }
          delete data[key];
          _this6._writeData(data, function (err2: any) {
            return cb(err2);
          });
        });
      })["catch"](cb);
    }
  }, {
    key: "_lockfileOptions",
    value: function _lockfileOptions() {
      return {
        realpath: false,
        retries: {
          retries: 5,
          minTimeout: 20
        }
      };
    }
  }, {
    key: "_releaseAndCb",
    value: function _releaseAndCb(release: any, cb: any) {
      return function (err: any) {
        if (err) {
          release().then(function () {
            return cb(err);
          })["catch"](function (releaseErr: any) {
            return cb(combineErrors([err, releaseErr]));
          });
          return;
        }
        release().then(cb)["catch"](cb);
      };
    }
  }, {
    key: "_writeData",
    value: function _writeData(data: any, cb: any) {
      var opts = {
        encoding: 'utf8',
        mode: 432,
        flag: 'w'
      };
      // @ts-expect-error TS(2345): Argument of type '{ encoding: string; mode: number... Remove this comment to see the full error message
      writeFile(this.path, JSON.stringify(data), opts, function (err) {
        return cb(err);
      });
    }
  }, {
    key: "_getData",
    value: function _getData(cb: any) {
      readFile(this.path, 'utf8', function (err, data) {
        if (err) {
          // return empty data if file does not exist
          if (err.code === 'ENOENT') cb(null, {});else cb(err);
          return;
        }
        try {
          data = !data.trim().length ? {} : JSON.parse(data);
        } catch (error) {
          cb(error);
          return;
        }
        cb(null, data);
      });
    }
  }]);
  return FileUrlStorage;
}();
